### PR TITLE
Mng 6957: Support Build Consumer while defining modules in reverse order

### PR DESCRIPTION
**Pull request by rfscholte**
Original summary: cannot build Maven Archetype with versionless reactor atifacts

### CODE REVIEW

**Comment by gnodet** [comment]
I think the problem raised in https://github.com/apache/maven/pull/402/commits/331e402b1a7f6e158444984d1272aeff18f54549 and https://github.com/apache/maven/pull/402/commits/aa285f20bc2c4aef9898e42e297dd43430ac2c11 are still valid with the proposed PR. Actually a bit worse since the ReactorModelCache now contains not only model data, but also a mapping from GA coordinates to Source.  Maybe this map could be set on the `DefaultTransformerContext`.  In all cases, the GA<->Source mapping is not really used as a cache (a cache should be optional and just avoids computing things multiple times) but as a storage (the `findRawModel` returns null the mapping has not been stored in the cache previously).

I'm willing to rework my work on top of yours, but the above point makes it much more difficult.  Could you consider revisit this point somehow ?

**Comment by gnodet** [comment]
> I think the problem raised in [331e402](https://github.com/apache/maven/commit/331e402b1a7f6e158444984d1272aeff18f54549) and [aa285f2](https://github.com/apache/maven/commit/aa285f20bc2c4aef9898e42e297dd43430ac2c11) are still valid with the proposed PR. Actually a bit worse since the ReactorModelCache now contains not only model data, but also a mapping from GA coordinates to Source. Maybe this map could be set on the `DefaultTransformerContext`. In all cases, the GA<->Source mapping is not really used as a cache (a cache should be optional and just avoids computing things multiple times) but as a storage (the `findRawModel` returns null the mapping has not been stored in the cache previously).
> 
> I'm willing to rework my work on top of yours, but the above point makes it much more difficult. Could you consider revisit this point somehow ?

I went ahead and did a small refactoring to avoid using the reactor cache, see https://github.com/gnodet/maven/commit/497e7febdcab42280b173da2c9454a1a159622e2

**Comment by michael-o** [comment]
I'd like to mention MNG-6983 which causes out of sync issues.

**Comment by michael-o** [comment]
I guess there is no way to break this up in a few logical PRs?

**Comment by rfscholte** [comment]
All is required to fix the issue, there's no clear way to split it.

**Comment by michael-o** [comment]
OK, thanks.

**Comment by gnodet** [comment]
@rfscholte overall, because of the number of commits, it's really difficult to comment on the diffs, maybe at this point, a squashed commit would be easier to review, since there are so many commits... Hence I'm raising a few points here rather than in possible diff comments...

### File vs Raw vs Effective model
The concept behind the raw model seems to change quite a bit.  I've watched the video, so I roughly get the idea.  However, there are a few spots where things are unclear : for example the `buildRawModel()` method from the `DefaultModelBuilder` seems to return... a file model.  It would be nice to have it documented somewhere.

### FileModelSource vs ArtifactModelSource
Could you please clarify the usage of `FileModelSource` ? There are a few spots where `modelSource instanceof FileModelSource` is used.  It seems that `FileModelSource` is used to denote a pom file which is being built (i.e. part of the reactor) whereas `ArtifactModelSource` denotes a pom file coming from a repository. What is unclear is why would is why the process would be specific to the Source implementation... If pom files should be read differently, shouldn't that be an explicit part of the `BuildModelRequest` instead of implicit from the type of the Source ?

### GA instead of GAV
All the model read seem to be indexed by group/artifact and not by group/artifact/version.  Does that mean that a single reactor can not used 2 different dependencies with different versions ? I know this can't happen for a single model, but this is currently supported for two different modules.  If the reactor only contains a single model for a given GA, how will that work ?

### SAX vs Stax
Another point is the usage of SAX rather than Stax API (honestly, I have hardly seen anybody using SAX since years). The models are read using a pull parser, so I'm not sure why introducing a push SAX filter to convert the file into a stream to be consumed by the pull parser in another thread.  This seems really inefficient.  In huge builds, a few hundreds or more pom files can be read, so that could really be a real problem.  I wonder if the `ModelSourceTransformer` would be better redesigned as a pull filter somehow.  Maybe something like
```
   XmlPullParser transform( XmlPullParser parser, TransformerContext context )
       throws IOException, TransformerException;
```
I know there are a bunch of things written in `maven-xml` already but this decision can have a big impact.  An alternative would be to only use a SAX parser, in which case the xml -> model mapping would have to be rewritten.  I don't really care, but it does not seem a very good option for the future mix push/pull parsing together to read a model.

### ModelReader
In the `DefaultModelReader`, only the `Model read( File input, Map<String, ?> options )` method does use the `ModelSourceTransformer` to read the Model.  If there's a good reason for not performing a transformation when using a `Reader` or an `InputStream`, it should be written in the javadoc, but it seems to me it can be extended to the other methods as well.

**Comment by rfscholte** [comment]
@gnodet thanks for reviewing. It doesn't make sense to review it per commit, I would suggest to comment on the total changes. In the end these commit will be squashed into on, yes.

File vs Raw vs Effective model
`ModelBuilder.buildRawModel( File pomFile, int validationLevel, boolean locationTracking )` was actually never used in our codebase. Now returning the rawModel only works if there is a fileModel. This is different compared to Maven3, where the fileModel and rawModel are actually the same.

FileModelSource vs ArtifactModelSource
Yes, artifactModelSources are loaded from the repository, FileModelSources are from the reactor/multimodule. The main reason is that pom from a repository should NEVER be transformed.

GA instead of GAV
GA is required for versionless dependencies, i.e. if there is only one GA in the project, now we can use that version. If for a weird reason there are multiple versions, an exception is thrown, because Maven cannot select a version: https://github.com/apache/maven/pull/391/files#diff-8af144c3b53554dd804a30126d20926a7f93aa03b86fdc12613a2f5c2df2b15cR57-R72

SAX vs Stax
I don't see the issues here. There is a need for chaining, hence the available XMLFilters made sense to me. For now I would like to stick to the current implementation, do an alpha-release and wait for feedback.

ModelReader
IIRC when I tried that several tests broke and I couldn't figure out how to solve that.

**Comment by gnodet** [comment]
@rfscholte I've pushed a better commit for the `ReactorModelCache` at https://github.com/gnodet/maven/commit/e60fce46da6c783724f9b552b7e491a8efba4037
There's stil the cast in `DefaultModelBuilder` (which could be transformed into an assertion given the `TransformerContextBuilder` is created by the `DefaultModelBuilder`, but the `TranformerContext` which is supposed to be immutable does not have the `putSource` method anymore.

**Comment by rfscholte** [comment]
@gnodet I've added commit, see https://github.com/apache/maven/pull/391/commits/50e8beb59eee74bcce8281ebc4ae40a90883539e

File: maven-core/src/main/java/org/apache/maven/project/DefaultProjectBuilder.java [hunk 1]
```diff
@@ -263,21 +264,14 @@ private DependencyResolutionResult resolveDependencies( MavenProject project, Re
...
-        return ids;
+        return profiles.stream().map( Profile::getId ).collect( Collectors.toList() );
     }
 
     private ModelBuildingRequest getModelBuildingRequest( InternalConfig config )
     {
         ProjectBuildingRequest configuration = config.request;
 
-        ModelBuildingRequest request = new DefaultModelBuildingRequest();
+        DefaultModelBuildingRequest request = new DefaultModelBuildingRequest();
```
Comment (michael-o) [inline maven-core/src/main/java/org/apache/maven/project/DefaultProjectBuilder.java:274]: This change is necessary to pass the builder?

File: maven-core/src/main/java/org/apache/maven/project/DefaultProjectBuilder.java [hunk 2]
```diff
@@ -386,44 +381,22 @@ private ModelSource createStubModelSource( Artifact artifact )
...
-                {
-                    return modelPool.get( groupId, artifactId, null );
-                }
-            };
-            request.getRepositorySession().getData().set( TransformerContext.KEY, context );
-        }
-
-        InternalConfig config = new InternalConfig( request, modelPool,
-                useGlobalModelCache() ? getModelCache() : new ReactorModelCache() );
+        InternalConfig config =
```
Comment (michael-o) [inline maven-core/src/main/java/org/apache/maven/project/DefaultProjectBuilder.java:384]: Please explain why the inner class with the feature flag is not necessary anymore.

File: maven-core/src/main/java/org/apache/maven/project/DefaultProjectBuilder.java [hunk 3]
```diff
@@ -386,44 +381,22 @@ private ModelSource createStubModelSource( Artifact artifact )
...
 
-        Map<String, MavenProject> projectIndex = new HashMap<>( 256 );
+        Map<File, MavenProject> projectIndex = new HashMap<>( 256 );
 
+        // phase 1: get file Models from the reactor.
         boolean noErrors =
             build( results, interimResults, projectIndex, pomFiles, new LinkedHashSet<>(), true, recursive,
                    config, poolBuilder );
-
+
```
Comment (michael-o) [inline maven-core/src/main/java/org/apache/maven/project/DefaultProjectBuilder.java:394]: Trailing WS

File: maven-core/src/main/java/org/apache/maven/project/DefaultProjectBuilder.java [hunk 4]
```diff
@@ -386,44 +381,22 @@ private ModelSource createStubModelSource( Artifact artifact )
...
         boolean noErrors =
             build( results, interimResults, projectIndex, pomFiles, new LinkedHashSet<>(), true, recursive,
                    config, poolBuilder );
-
+        
         ClassLoader oldContextClassLoader = Thread.currentThread().getContextClassLoader();
 
         try
         {
+            // Phase 2: get effective from the reactor
```
Comment (michael-o) [inline maven-core/src/main/java/org/apache/maven/project/DefaultProjectBuilder.java:399]: get effective what?

File: maven-core/src/main/java/org/apache/maven/project/DefaultProjectBuilder.java [hunk 5]
```diff
@@ -505,32 +483,17 @@ private boolean build( List<ProjectBuildingResult> results, List<InterimResult>
...
-                  iarte ) );
-        }
-
-        projectIndex.put( result.getModelIds().get( 0 ), project );
-
         InterimResult interimResult = new InterimResult( pomFile, request, result, listener, isRoot );
         interimResults.add( interimResult );
-
-        if ( recursive && !model.getModules().isEmpty() )
+
```
Comment (michael-o) [inline maven-core/src/main/java/org/apache/maven/project/DefaultProjectBuilder.java:492]: Trailing WS

File: maven-core/src/main/java/org/apache/maven/project/DefaultProjectBuilder.java [hunk 6]
```diff
@@ -640,12 +605,12 @@ private boolean build( List<ProjectBuildingResult> results, List<InterimResult>
...
 
     private boolean build( List<ProjectBuildingResult> results, List<MavenProject> projects,
-                           Map<String, MavenProject> projectIndex, List<InterimResult> interimResults,
+                           Map<File, MavenProject> projectIndex, List<InterimResult> interimResults,
                            ProjectBuildingRequest request, Map<File, Boolean> profilesXmls,
                            RepositorySystemSession session )
     {
         boolean noErrors = true;
-
+
```
Comment (michael-o) [inline maven-core/src/main/java/org/apache/maven/project/DefaultProjectBuilder.java:613]: Trailing WS

File: maven-core/src/main/java/org/apache/maven/project/DefaultProjectBuilder.java [hunk 7]
```diff
@@ -685,12 +650,15 @@ private boolean build( List<ProjectBuildingResult> results, List<MavenProject> p
...
             {
                 DefaultProjectBuildingResult result = null;
-                if ( project == null )
+                if ( project == null || interimResult.result.getEffectiveModel() == null )
                 {
                     result = new DefaultProjectBuildingResult( e.getModelId(), interimResult.pomFile, e.getProblems() );
                 }
                 else
                 {
+                    // building effective model includes validation, make sure effectiveModel is available
```
Comment (michael-o) [inline maven-core/src/main/java/org/apache/maven/project/DefaultProjectBuilder.java:659]: use consistent comment on effective model

File: maven-core/src/main/java/org/apache/maven/project/DefaultProjectBuilder.java [hunk 8]
```diff
@@ -703,15 +671,14 @@ private boolean build( List<ProjectBuildingResult> results, List<MavenProject> p
...
-    private void initProject( MavenProject project, Map<String, MavenProject> projects,
+    private void initProject( MavenProject project, Map<File, MavenProject> projects,
                               boolean buildParentIfNotExisting, ModelBuildingResult result,
                               Map<File, Boolean> profilesXmls, ProjectBuildingRequest projectBuildingRequest )
     {
         Model model = result.getEffectiveModel();
 
         project.setModel( model );
-        project.setOriginalModel( result.getRawModel() );
-        project.setFile( model.getPomFile() );
```
Comment (michael-o) [inline maven-core/src/main/java/org/apache/maven/project/DefaultProjectBuilder.java:714]: Why is this not required anymore?

File: maven-core/src/main/java/org/apache/maven/project/ReactorModelCache.java [hunk 9]
```diff
@@ -36,7 +36,7 @@
 {
 
     private final Map<Object, Object> models = new ConcurrentHashMap<>( 256 );
-
+
```
Comment (michael-o) [inline maven-core/src/main/java/org/apache/maven/project/ReactorModelCache.java:39]: Trailing WS

File: maven-model-builder/src/main/java/org/apache/maven/model/building/BuildModelSourceTransformer.java [hunk 10]
```diff
@@ -58,8 +58,8 @@ protected AbstractSAXFilter getSAXFilter( Path pomFile,
         throws TransformerConfigurationException, SAXException, ParserConfigurationException
     {
         BuildPomXMLFilterFactory buildPomXMLFilterFactory =
-            new DefaultBuildPomXMLFilterFactory( context, lexicalHandlerConsumer );
-
+            new DefaultBuildPomXMLFilterFactory( context, lexicalHandlerConsumer, false );
+
```
Comment (michael-o) [inline maven-model-builder/src/main/java/org/apache/maven/model/building/BuildModelSourceTransformer.java:62]: Trailing WS

File: maven-model-builder/src/main/java/org/apache/maven/model/building/DefaultBuildPomXMLFilterFactory.java [hunk 11]
```diff
@@ -32,26 +32,29 @@
...
  */
 public class DefaultBuildPomXMLFilterFactory extends BuildPomXMLFilterFactory
 {
     private final TransformerContext context;
     
+    /**
+     *
+     * @param context a set of data to extract values from as required for the build pom
+     * @param lexicalHandlerConsumer the lexical handler consumer
+     * @param consume true is this factory used for creating the consumer pom, otherwise false
```
Comment (michael-o) [inline maven-model-builder/src/main/java/org/apache/maven/model/building/DefaultBuildPomXMLFilterFactory.java:48]: is => if. Use @code for boolean literals

File: maven-model-builder/src/main/java/org/apache/maven/model/building/DefaultModelBuilder.java [hunk 12]
```diff
@@ -114,7 +120,7 @@
 
     @Inject
     private ModelPathTranslator modelPathTranslator;
-
+
```
Comment (michael-o) [inline maven-model-builder/src/main/java/org/apache/maven/model/building/DefaultModelBuilder.java:123]: Trailing WS

File: maven-model-builder/src/main/java/org/apache/maven/model/building/DefaultModelBuilder.java [hunk 13]
```diff
@@ -649,64 +682,101 @@ private Model readModel( ModelSource modelSource, File pomFile, ModelBuildingReq
...
         intoCache( request.getModelCache(), groupId, artifactId, version, ModelCacheTag.RAW, modelData );
-
-        return model;
+        intoCache( request.getModelCache(), modelSource, ModelCacheTag.RAW, modelData );
+ 
+        return rawModel;
     }
 
-    private Model getModelFromCache( ModelSource modelSource, ModelCache cache )
+    private Model getModelFromCache( Source modelSource, ModelCache cache )
```
Comment (michael-o) [inline maven-model-builder/src/main/java/org/apache/maven/model/building/DefaultModelBuilder.java:779]: Should the arg name change too?

File: maven-model-builder/src/main/java/org/apache/maven/model/building/DefaultModelBuilder.java [hunk 14]
```diff
@@ -1707,4 +1786,131 @@ protected void mergePluginContainer_Plugins( PluginContainer target, PluginConta
...
         }
     }
+    
+    /**
+     * Builds up the transformer context.
+     * After the buildplan is ready, the build()-method returns the immutable context useful during distribution.
+     * This is an inner class, as it must be able to call readRawModel() 
+     * 
+     * @author Robert Scholte
+     * @since 3.7.0
```
Comment (michael-o) [inline maven-model-builder/src/main/java/org/apache/maven/model/building/DefaultModelBuilder.java:1796]: not valid anymore

File: maven-model-builder/src/main/java/org/apache/maven/model/building/DefaultModelBuilder.java [hunk 15]
```diff
@@ -1707,4 +1786,131 @@ protected void mergePluginContainer_Plugins( PluginContainer target, PluginConta
...
+                        }
+                    }
+                    return null;
+                }
+                
+                private Model findRawModel( Path p )
+                {
+                    if ( !Files.isRegularFile( p ) )
+                    {
+                        throw new IllegalArgumentException( "Not a regular file" + p );
```
Comment (michael-o) [inline maven-model-builder/src/main/java/org/apache/maven/model/building/DefaultModelBuilder.java:1869]: Makes it either with a colon or use `'{}'`

File: maven-model-builder/src/main/java/org/apache/maven/model/building/DefaultModelBuilder.java [hunk 16]
```diff
@@ -1707,4 +1786,131 @@ protected void mergePluginContainer_Plugins( PluginContainer target, PluginConta
...
+        public Source getSource( String groupId, String artifactId )
+        {
+            Set<Source> sources = mappedSources.get( new DefaultTransformerContext.GAKey( groupId, artifactId ) );
+            if ( sources == null )
+            {
+                return null;
+            }
+            return sources.stream().reduce( ( a, b ) ->
+            {
+                throw new IllegalStateException( "No unique Source for " + groupId + ':' + artifactId
```
Comment (michael-o) [inline maven-model-builder/src/main/java/org/apache/maven/model/building/DefaultModelBuilder.java:1904]: String format?

File: maven-model-builder/src/main/java/org/apache/maven/model/building/DefaultModelBuildingRequest.java [hunk 17]
```diff
@@ -386,28 +385,27 @@ public DefaultModelBuildingRequest setModelCache( ModelCache modelCache )
     }
 
     @Override
-    public Model getFileModel()
+    public Model getRawModel()
```
Comment (michael-o) [inline maven-model-builder/src/main/java/org/apache/maven/model/building/DefaultModelBuildingRequest.java:388]: Why did you change position of these methods?

File: maven-model-builder/src/main/java/org/apache/maven/model/building/DefaultTransformerContext.java [hunk 18]
```diff
@@ -0,0 +1,90 @@
...
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Objects;
+
+import org.apache.maven.model.Model;
+
+/**
+ * 
+ * @author Robert Scholte
+ * @since 3.7.0
```
Comment (michael-o) [inline maven-model-builder/src/main/java/org/apache/maven/model/building/DefaultTransformerContext.java:32]: Not valid anymore

File: maven-model-builder/src/main/java/org/apache/maven/model/building/DefaultTransformerContext.java [hunk 19]
```diff
@@ -0,0 +1,90 @@
...
+        }
+
+        @Override
+        public boolean equals( Object obj )
+        {
+            if ( this == obj )
+            {
+                return true;
+            }
+            GAKey other = (GAKey) obj;
```
Comment (michael-o) [inline maven-model-builder/src/main/java/org/apache/maven/model/building/DefaultTransformerContext.java:86]: What about instance of?

File: maven-model-builder/src/main/java/org/apache/maven/model/building/FilterModelBuildingRequest.java [hunk 20]
```diff
@@ -255,30 +255,30 @@ public FilterModelBuildingRequest setModelCache( ModelCache modelCache )
 
         return this;
     }
-
+
```
Comment (michael-o) [inline maven-model-builder/src/main/java/org/apache/maven/model/building/FilterModelBuildingRequest.java:258]: Trailing WS

File: maven-model-builder/src/main/java/org/apache/maven/model/building/FilterModelBuildingRequest.java [hunk 21]
```diff
@@ -255,30 +255,30 @@ public FilterModelBuildingRequest setModelCache( ModelCache modelCache )
 
         return this;
     }
-
+    
     @Override
-    public Model getFileModel()
+    public Model getRawModel()
```
Comment (michael-o) [inline maven-model-builder/src/main/java/org/apache/maven/model/building/FilterModelBuildingRequest.java:260]: Same here

File: maven-model-builder/src/main/java/org/apache/maven/model/building/ModelBuildingRequest.java [hunk 22]
```diff
@@ -63,34 +63,35 @@
      * Denotes strict validation as recommended by the current Maven version.
      */
     int VALIDATION_LEVEL_STRICT = VALIDATION_LEVEL_MAVEN_3_0;
-    
+
     /**
-     * 
-     * @return the file model
+     * Gets the file model to build (with profile activatio).
```
Comment (michael-o) [inline maven-model-builder/src/main/java/org/apache/maven/model/building/ModelBuildingRequest.java:68]: Typo

File: maven-model-builder/src/main/java/org/apache/maven/model/building/ModelBuildingRequest.java [hunk 23]
```diff
@@ -63,34 +63,35 @@
...
-     * 
+     * Set the file model with profile activation
+     *
      * @param fileModel
      * @return This request, never {@code null}.
      * @since 4.0.0
      */
     ModelBuildingRequest setFileModel( Model fileModel );
-
+
```
Comment (michael-o) [inline maven-model-builder/src/main/java/org/apache/maven/model/building/ModelBuildingRequest.java:84]: Trailing WS

File: maven-model-builder/src/main/java/org/apache/maven/model/building/ModelBuildingRequest.java [hunk 24]
```diff
@@ -63,34 +63,35 @@
...
      * @since 4.0.0
      */
     ModelBuildingRequest setFileModel( Model fileModel );
-
+    
     /**
-     * Gets the raw model to build. If not set, model source will be used to load raw model.
-     *
-     * @return The raw model to build or {@code null} if not set.
+     * @deprecated rawModel is never set, instead the fileModel is set
```
Comment (michael-o) [inline maven-model-builder/src/main/java/org/apache/maven/model/building/ModelBuildingRequest.java:86]: Why deprecate in a major version?

File: maven-model-builder/src/main/java/org/apache/maven/model/building/ModelBuildingResult.java [hunk 25]
```diff
@@ -41,6 +41,13 @@
      * @return The model identifiers from the lineage of models, never {@code null}.
      */
     List<String> getModelIds();
+    
+    /**
+     * 
+     * @return the file model
+     * @since 3.7.0
```
Comment (michael-o) [inline maven-model-builder/src/main/java/org/apache/maven/model/building/ModelBuildingResult.java:48]: Not valid anymore

File: maven-model-builder/src/main/java/org/apache/maven/model/building/ModelCacheTag.java [hunk 26]
```diff
@@ -129,12 +129,16 @@ public DependencyManagement fromCache( DependencyManagement data )
 
     };
 
-    ModelCacheTag<Model> FILEMODEL = new ModelCacheTag<Model>() 
+    /**
+     * The tag used for the file model without profile activation
+     * @since 3.7.0
```
Comment (michael-o) [inline maven-model-builder/src/main/java/org/apache/maven/model/building/ModelCacheTag.java:134]: Not valid anymore

File: maven-model-builder/src/main/java/org/apache/maven/model/building/TransformerContextBuilder.java [hunk 27]
```diff
@@ -0,0 +1,46 @@
...
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+/**
+ * The transformerContextBuilder is responsible for initializing the TransformerContext.
+ * In case rawModels are missing, it could do new buildingRequests on the ModelBuilder.
+ * 
+ * @author Robert Scholte
+ * @since 3.7.0
```
Comment (michael-o) [inline maven-model-builder/src/main/java/org/apache/maven/model/building/TransformerContextBuilder.java:27]: Not valid anymore

File: maven-model-builder/src/main/java/org/apache/maven/model/io/ModelReader.java [hunk 28]
```diff
@@ -47,6 +47,12 @@
      * location tracking.
      */
     String INPUT_SOURCE = "org.apache.maven.model.io.inputSource";
+    
+    /**
+     * The key for the option to provide a transformer context, which can be used to transform the input while reading
+     * to get an advanced version of the model.
+     */
+    String TRANSFORMER_CONTEXT = "transformerContext";
```
Comment (michael-o) [inline maven-model-builder/src/main/java/org/apache/maven/model/io/ModelReader.java:55]: Interesting that those aren't `static final`

File: maven-xml/src/test/java/org/apache/maven/xml/sax/filter/ParentXMLFilterTest.java [hunk 29]
```diff
@@ -89,6 +89,31 @@ public void testDefaultRelativePath() throws Exception
...
+        String input = "<parent>"
+            + "<groupId>GROUPID</groupId>"
+            + "<artifactId>ARTIFACTID</artifactId>"
+            + "<relativePath></relativePath>"
+            + "</parent>";
+        String expected = "<parent>"
+                        + "<groupId>GROUPID</groupId>"
+                        + "<artifactId>ARTIFACTID</artifactId>"
+                        + "<relativePath/>" // SAX optimization, however "" != null ...
+                        + "</parent>";;
```
Comment (michael-o) [inline maven-xml/src/test/java/org/apache/maven/xml/sax/filter/ParentXMLFilterTest.java:110]: Redundant semicolon

**Review by michael-o — CHANGES_REQUESTED** [review]

File: maven-core/src/main/java/org/apache/maven/project/DefaultProjectBuilder.java (hunk 1 above)
Comment (rfscholte) [reply maven-core/src/main/java/org/apache/maven/project/DefaultProjectBuilder.java]: Not anymore, during refactoring there was a period where I needed it. Changed it back.

File: maven-core/src/main/java/org/apache/maven/project/DefaultProjectBuilder.java (hunk 3 above)
Comment (rfscholte) [reply maven-core/src/main/java/org/apache/maven/project/DefaultProjectBuilder.java]: Fixed

File: maven-core/src/main/java/org/apache/maven/project/DefaultProjectBuilder.java (hunk 4 above)
Comment (rfscholte) [reply maven-core/src/main/java/org/apache/maven/project/DefaultProjectBuilder.java]: models, fixed comment

File: maven-core/src/main/java/org/apache/maven/project/DefaultProjectBuilder.java (hunk 5 above)
Comment (rfscholte) [reply maven-core/src/main/java/org/apache/maven/project/DefaultProjectBuilder.java]: fixed

File: maven-model-builder/src/main/java/org/apache/maven/model/building/ModelBuildingResult.java (hunk 25 above)
Comment (rfscholte) [reply maven-model-builder/src/main/java/org/apache/maven/model/building/ModelBuildingResult.java]: fixed

File: maven-model-builder/src/main/java/org/apache/maven/model/building/ModelCacheTag.java (hunk 26 above)
Comment (rfscholte) [reply maven-model-builder/src/main/java/org/apache/maven/model/building/ModelCacheTag.java]: fixed

File: maven-model-builder/src/main/java/org/apache/maven/model/building/TransformerContextBuilder.java (hunk 27 above)
Comment (rfscholte) [reply maven-model-builder/src/main/java/org/apache/maven/model/building/TransformerContextBuilder.java]: fixed

File: maven-xml/src/test/java/org/apache/maven/xml/sax/filter/ParentXMLFilterTest.java (hunk 29 above)
Comment (rfscholte) [reply maven-xml/src/test/java/org/apache/maven/xml/sax/filter/ParentXMLFilterTest.java]: fixed

File: maven-core/src/main/java/org/apache/maven/project/DefaultProjectBuilder.java (hunk 8 above)
Comment (rfscholte) [reply maven-core/src/main/java/org/apache/maven/project/DefaultProjectBuilder.java:714]: File was already set on the project when creating it (line 167)
Also, the Model was being abused to hold the file. A model is created with an InputStream, so the file information was added via post-processing. Only the MavenProject is interested in the file, not the Model.

File: maven-model-builder/src/main/java/org/apache/maven/model/building/DefaultModelBuilder.java [hunk 30]
```diff
@@ -1901,8 +1901,9 @@ public Source getSource( String groupId, String artifactId )
             }
             return sources.stream().reduce( ( a, b ) ->
             {
-                throw new IllegalStateException( "No unique Source for " + groupId + ':' + artifactId
-                        + ": " + a.getLocation() + " and " + b.getLocation() );
+                throw new IllegalStateException( String.format( "No unique Source for %s:%s: %s and %s ",
```
Comment (michael-o) [inline maven-model-builder/src/main/java/org/apache/maven/model/building/DefaultModelBuilder.java:1904]: There is a trailing WS in the string

File: maven-model-builder/src/main/java/org/apache/maven/model/building/DefaultModelBuilder.java [hunk 31]
```diff
@@ -1707,4 +1786,132 @@ protected void mergePluginContainer_Plugins( PluginContainer target, PluginConta
...
+        }
+
+        public Source getSource( String groupId, String artifactId )
+        {
+            Set<Source> sources = mappedSources.get( new DefaultTransformerContext.GAKey( groupId, artifactId ) );
+            if ( sources == null )
+            {
+                return null;
+            }
+            return sources.stream().reduce( ( a, b ) ->
```
Comment (pzygielo) [inline maven-model-builder/src/main/java/org/apache/maven/model/building/DefaultModelBuilder.java:1902]: I don't understand. Is the purpose of this to
```java
if (sources.size() > 1)
  return null;
else
  throw new ISEx();
```

**Comment by michael-o** [comment]
@rfscholte Please also rebase/merge master into this branch. I'd like to run all ITs with that.

File: maven-model-builder/src/main/java/org/apache/maven/model/building/DefaultModelBuilder.java (hunk 31 above)
Comment (rfscholte) [reply maven-model-builder/src/main/java/org/apache/maven/model/building/DefaultModelBuilder.java:1902]: no, effectively this does:

    if (sources.size() == 0)
      return null;
    else if (sources.size() == 1)
        return sources.get(0);
    else
       throw new ISEx();

reduce is a only used when there is something to reduce. In other words, there should be at least  2 items.

File: maven-model-builder/src/main/java/org/apache/maven/model/building/ModelBuildingRequest.java (hunk 24 above)
Comment (rfscholte) [reply maven-model-builder/src/main/java/org/apache/maven/model/building/ModelBuildingRequest.java:86]: I don't know if it is used by projects like polyglot and tiles. ModelBuildingRequest is used in an API.
Removing it now might imply that those projects can only be used with Maven3 or Maven4, not both. I don't think we should do that.

File: maven-model-builder/src/main/java/org/apache/maven/model/building/DefaultModelBuilder.java (hunk 31 above)
Comment (pzygielo) [reply maven-model-builder/src/main/java/org/apache/maven/model/building/DefaultModelBuilder.java:1902]: I wrote not what I had meant (still my mistake). But you explained it anyway :grin: and now I see it - thanks.

File: maven-core/src/main/java/org/apache/maven/project/ReactorModelCache.java [hunk 32]
```diff
@@ -113,7 +107,48 @@ public int hashCode()
...
+    {
+        private final String groupId;
+        
+        private final String artifactId;
+        
+        private final int hashCode;
 
+        GACacheKey( String groupId, String artifactId )
+        {
+            super();
```
Comment (michael-o) [inline maven-core/src/main/java/org/apache/maven/project/ReactorModelCache.java:122]: This super is is redundant, isn't it?

**Review by michael-o — COMMENTED** [review]
One more nit, but I cannot object anything here anymore. Going through ITs now.

File: maven-core/src/main/java/org/apache/maven/project/ReactorModelCache.java (hunk 32 above)
Comment (rfscholte) [reply maven-core/src/main/java/org/apache/maven/project/ReactorModelCache.java:122]: yes, it is there already implicitly, but I'll remove it.

**Review by michael-o — APPROVED** [review]
LGTM now. Please squash along with ITs, I will run again.

**Comment by jira-importer** [comment]
Resolve #9731
